Purge resources and fix FedoraLdpTests

whikloj opened the pull request:
**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3034

# What does this Pull Request do?
Allows you to DELETE a `fcr:tombstone` to remove the sidecar files (if part of an Archival Group) or delete the entire object.

Also re-enables the FedoraLdpTests unit test class and fixes all the tests.

# How should this be tested?

Build the PR.
Create an object.
Find the object in your storage root.
Delete the object, see that the object still exists but the sidecar file has the `deleted: true` property.
Delete the tombstone, see the entire OCFL object is removed.

# Interested parties
@fcrepo4/committers

awoods commented on fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java:
```diff
@@ -79,13 +82,14 @@
      */
     private static final String RESOURCES_TABLE_DDL = "CREATE TABLE IF NOT EXISTS " + RESOURCES_TABLE + " (" +
```
Comment for this statement should be updated to include 'deleted' info.

awoods commented:
The basic functionality of purge works exactly as expected. However, in testing purge functionality, the following issues were surfaced. Most of these issues are likely not a direct result of this pull-request.

### Tombstone response code
- When performing a `GET` request on an `/fcr:tombstone`, a 404 is returned. 
- Fedora 5.1.0 returns a "405 Method Not Allowed", which seems reasonable
```
curl -i -XPUT -H"Content-Type: text/turtle" -d "<> <http://example.org#title> 'a title'" localhost:8080/rest/container
curl -i localhost:8080/rest/container
curl -i -XDELETE localhost:8080/rest/container
curl -i localhost:8080/rest/container/fcr:tombstone
```

### Purge leaves invalid OCFL storage hierarchy
- After creating, deleting and purging an atomic resource, the path to the purged OCFL object remains, which is invalid.
```
curl -i -XPUT -H"Content-Type: text/turtle" -d "<> <http://example.org#title> 'a title'" localhost:8080/rest/container
curl -i localhost:8080/rest/container
find /tmp/fcrepo.ocfl.storage.root.dir -name container
curl -i -XDELETE localhost:8080/rest/container
curl -i localhost:8080/rest/container/fcr:tombstone
curl -i -XDELETE localhost:8080/rest/container/fcr:tombstone
// Inspect path returned in the "find" command above... notice dangling directory
```

### Memento fo deleted resource still available
- If a resource is deleted, access to its mementos is still available... I would expect to get a tombstone instead
```
curl -i -XPUT -d "some text" localhost:8080/rest/binary
curl -i localhost:8080/rest/binary/fcr:versions
curl -i localhost:8080/rest/binary/fcr:versions/<memento-id>
curl -i -XDELETE localhost:8080/rest/binary
// GET memento of deleted resource... still works??
curl -i localhost:8080/rest/binary/fcr:versions/<memento-id> 
```

### Reincarnated children
- Rebuild of containers with deleted children as well as archival groups with deleted children result in reincarnated children upon Fedora rebuild
```
curl -i -XPUT -H'Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"' localhost:8080/rest/ag
curl -i localhost:8080/rest/ag
curl -i -XPUT -d "a file1" localhost:8080/rest/ag/file1
curl -i localhost:8080/rest/ag/file1
curl -i -XDELETE localhost:8080/rest/ag/file1
curl -i localhost:8080/rest/ag/file1
// Stop Fedora, delete work-dir, start Fedora
curl -i localhost:8080/rest/ag
// Notice the child "file1" is ldp:contains by "ag" - BAD
```

### Purged grandchild cause trouble
- Create an archival group with two levels of nesting: child/grandchild
- Delete the child - GOOD
- Purge the grandchild - GOOD
- Purge the child - BOOM!
- Additionally, try to purge the archival group - DOUBLE BOOM!
```
curl -i -XPUT -H'Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"' localhost:8080/rest/ag
curl -i localhost:8080/rest/ag
curl -i -XPUT  localhost:8080/rest/ag/child
curl -i -XPUT -d "a file" localhost:8080/rest/ag/child/grandchild
curl -i -XDELETE localhost:8080/rest/ag/child
curl -i -XDELETE localhost:8080/rest/ag/child/grandchild/fcr:tombstone
curl -i -XDELETE localhost:8080/rest/ag/child/fcr:tombstone
// Notice above request results in an exception - BAD
curl -i -XDELETE localhost:8080/rest/ag/fcr:tombstone
// Notice above request results in another exception - BAD
```

### Deleting resource in a transaction
- When creating and deleting a resource within the same transaction... BOOM!
- This is true for both containers and binaries
```
curl -i -X POST "http://localhost:8080/rest/fcr:tx"
txid=<txn-URI>
// Buy ourselves some extra time
curl -i -XPOST ${txid}
curl -i -H"Atomic-ID: ${txid}" -XPUT -d "some text" localhost:8080/rest/binary
curl -i -H"Atomic-ID: ${txid}"  localhost:8080/rest/binary
curl -i -H"Atomic-ID: ${txid}"  -XDELETE localhost:8080/rest/binary
// Notice above request results in stacktrace - BAD
```

pwinckles commented:
@awoods Your second point, "Purge leaves invalid OCFL storage hierarchy" is a known bug in ocfl-java that I haven't gotten around to fixing yet.

https://github.com/UW-Madison-Library/ocfl-java/issues/8

whikloj commented on fcrepo-http-api/src/test/java/org/fcrepo/http/api/FedoraLdpTest.java:
```diff
@@ -1007,68 +1059,93 @@ public void testPatchBinary() throws MalformedRdfException, IOException {
         assertEquals(CREATED.getStatusCode(), actual.getStatus());
     }
 
+    @Ignore("HttpRdfService cannot parse sparql-update")
```
No I didn't, this related to my question if we wanted to continue to support PUT/POSTting RDF using the application/sparql-update syntax. The JIRA should either ensure that we properly support or properly reject sparql on PUT/POST. Just not sure which one yet.

awoods replied:
I would suggest rejecting sparql-update on PUT/POST.

whikloj replied:
Would a PUT with application/sparql-update now become a binary containing the body of the message or are we excluding that type entirely?

whikloj commented:
The commit addresses the comments except around a JIRA for PUTting sparql-update. It also makes mementos for deleted resources are not accessible and that tombstones can only be addressed with DELETE requests.

awoods commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java:
```diff
@@ -1087,7 +1086,14 @@ public FedoraResource getResourceFromPath(final String externalPath) {
+            }
+
+            if (originalResource instanceof Tombstone) {
                 final String tombstoneUri = identifierConverter().toExternalId(
```
I was hoping that the test entitled: "Memento of deleted resource still available" would work given this change. However, following those test steps results in:
```
HTTP/1.1 400 Bad Request
Path is invalid: /binary/fcr:versions/20200603235900/fcr:tombstone
```

whikloj replied:
This should now be resolved. Though the test that confirms this might need some work as it was passing before the most recent change so I'm not sure it is actually useful.

awoods commented:
Noting that there will be some merge conflicts with:
https://github.com/fcrepo4/fcrepo4/pull/1689/files#diff-82746a31e1e2e17f71ac18a3e856339b

whikloj commented on fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java:
```diff
@@ -280,69 +391,75 @@ public String getContainedBy(final String txID, final FedoraId resource) {
 
     @Override
-    public void removeResource(final String txID, final FedoraId resource) {
+    public void purgeContainedBy(@Nonnull final String txID, final FedoraId parent, final FedoraId child) {
```
This method no longer exists, as I needed to use the other one (`purgeResource()`).

But in either case a record is added to the `transaction_operations` table for the purge and then those changes are committed to the actual `resource` table inside a database transaction as part of the `commit()`

pwinckles replied:
It is probably a small enough issue that it doesn't merit worrying about too much, but, in `purgeResource`, it is technically possible for the rollback of a delete to be applied but the purge not be applied. In that case, the `transaction_operations` table is in an incorrect state and it doesn't matter that it's committed to the `resources` table in a transaction.

whikloj replied:
Fair point, I'm starting to hate this table and think Fedora should support 1 connection only. I mean how many people _really_ need to work with the repository?

whikloj commented:
To clarify the issues @awoods mentioned above
* Tombstone response code - is resolved in this PR
* Purge leaves invalid OCFL storage hierarchy - is related to the ocfl-java client leaving the pairtree directory structure but properly removing the OCFL object directory. (https://github.com/UW-Madison-Library/ocfl-java/issues/8)
* Memento of deleted resource still available - is resolved in this PR
* Reincarnated children - https://jira.lyrasis.org/browse/FCREPO-3338
* Create/delete in same transaction - https://jira.lyrasis.org/browse/FCREPO-3340
* Purged grandchild cause trouble - Resolved in this PR and IT added.
